go/nomdl/parser.go

```diff
@@ -0,0 +1,224 @@
+// is an error.
+func MustParseType(code string) *types.Type {
+	typ, err := ParseType(code)
+	d.PanicIfError(err)
+	return typ
+}
+
+func (p *Parser) ensureAtEnd() {
+	p.lex.eat(scanner.EOF)
+}
+
+// Type :
+//   TypeWithoutUnion (`|` TypeWithoutUnion)*
+//
+// TypeWithoutUnion :
+//   `Blob`
+//   `Bool`
+//   `Number`
+//   `String`
+//   `Type`
+//   `Value`
+//   CycleType
+//   ListType
+//   MapType
+//   RefType
+//   SetType
+//   StructType
+//
+// CycleType :
+//   `Cycle` `<` uint32 `>`
+//
+// ListType :
+//   `List` `<` Type? `>`
+//
+// MapType :
+//   `Map` `<` (Type `,` Type)? `>`
+//
+// RefType :
+//   `Set` `<` Type `>`
+//
+// SetType :
+//   `Set` `<` Type? `>`
+//
+// StructType :
+//   `struct` StructName? `{` StructField? | (StructField `,`)* `}`
+//
+// StructName :
+//   Ident
+//
+// StructField :
+//   StructFieldName `:` Type
+//
+// StructFieldName :
+//   Ident
+
+func (p *Parser) parseType() *types.Type {
+	t := p.parseTypeWithoutUnion()
+	tok := p.lex.peek()
+	if tok.Type != '|' {
+		return t
+	}
+	unionTypes := []*types.Type{t}
+	for p.lex.eatIf('|') {
+		unionTypes = append(unionTypes, p.parseTypeWithoutUnion())
+	}
+	return types.MakeUnionType(unionTypes...)
+}
+
+func (p *Parser) parseTypeWithoutUnion() *types.Type {
+	tok := p.lex.next()
+	switch tok.Type {
+	case '(':
+		panic("not implemented")
+	case scanner.Ident:
+		switch tok.Text {
+		case "Bool":
+			return types.BoolType
+		case "Blob":
+			return types.BlobType
+		case "Number":
+			return types.NumberType
+		case "String":
+			return types.StringType
+		case "Type":
+			return types.TypeType
+		case "Value":
+			return types.ValueType
+		case "struct":
+			return p.parseStructType()
+		case "Map":
+			return p.parseMapType()
+		case "List":
+			elemType := p.parseSingleElemType(true)
+			return types.MakeListType(elemType)
+		case "Set":
+			elemType := p.parseSingleElemType(true)
+			return types.MakeSetType(elemType)
+		case "Ref":
+			elemType := p.parseSingleElemType(false)
+			return types.MakeRefType(elemType)
+		case "Cycle":
+			return p.parseCycleType()
+		}
+	}
+	p.lex.unexpectedToken(tok)
+	return nil
+}
+
+func (p *Parser) parseStructType() *types.Type {
+	tok := p.lex.next()
+	name := ""
+	if tok.Type == scanner.Ident {
+		name = tok.Text
+		p.lex.eat('{')
+	} else {
+		p.lex.check('{', tok)
+	}
+	fields := types.FieldMap{}
+	for {
+		tok := p.lex.next()
+		if tok.Type == '}' {
+			break
+		}
+		if tok.Type == scanner.Ident {
+			fieldName := tok.Text
+			p.lex.eat(':')
+			typ := p.parseType()
+			fields[fieldName] = typ
+
+			if p.lex.eatIf('}') {
+				break
+			}
+			p.lex.eat(',')
+		} else {
+			p.lex.tokenMismatch('}', tok)
+		}
+	}
+	return types.MakeStructTypeFromFields(name, fields)
+}
+
+func (p *Parser) parseSingleElemType(allowEmptyUnion bool) *types.Type {
+	p.lex.eat('<')
+	if allowEmptyUnion && p.lex.eatIf('>') {
+		return types.MakeUnionType()
+	}
```

Comment:
Interesting...this `allowEmptyUnion` means that `List<>` and `Set<>` are valid? What do those mean in practical Noms?